# Carregando arquivo de dados

In [1]:
with open('corretor-master/artigos.txt', 'r', errors = 'ignore', encoding = 'utf-8') as f:
    artigos = f.read()

In [2]:
print(artigos[:500])




imagem 

Temos a seguinte classe que representa um usuário no nosso sistema:

java

Para salvar um novo usuário, várias validações são feitas, como por exemplo: Ver se o nome só contém letras, [**o CPF só números**] e ver se o usuário possui no mínimo 18 anos. Veja o método que faz essa validação:

java 

Suponha agora que eu tenha outra classe, a classe `Produto`, que contém um atributo nome e eu quero fazer a mesma validação que fiz para o nome do usuário: Ver se só contém letras. E aí? Vou


In [3]:
len(artigos)

2605046

In [4]:
texto_exemplo = "Olá, tudo bem?"
tokens = texto_exemplo.split()

In [5]:
tokens

['Olá,', 'tudo', 'bem?']

In [6]:
len(tokens)

3

# NLTK separando os tokens

In [7]:
import nltk

In [8]:
#nltk.download('punkt')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\Dorivan\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [9]:
# Separa as palavras dos caracteres
palavras_separadas = nltk.tokenize.word_tokenize(texto_exemplo)

In [10]:
palavras_separadas

['Olá', ',', 'tudo', 'bem', '?']

In [11]:
len(palavras_separadas)

5

In [12]:
'palavra'.isalpha()

True

In [13]:
# Percorre todos os dados do parâmetro, e atribui a uma lista apenas as palavras que são Alpha, que conteém letras
def separa_palavras(lista_tokens):
    lista_palavras = []
    for token in lista_tokens:
        if token.isalpha():
            lista_palavras.append(token)
    return lista_palavras

In [14]:
# Testando a função
separa_palavras(palavras_separadas)

['Olá', 'tudo', 'bem']

In [15]:
# Executando a função com os dados do 'artigo'
# variável que recebe as funcões de tokenização da base 'artigos'
lista_tokens = nltk.tokenize.word_tokenize(artigos)

# Executa a função com a variável que recebeu a tokenização
lista_palavras = separa_palavras(lista_tokens)

# Exibe a quantidade de palavras encontradas apenas com letras na base 'artigos'
print('O número de palavras é', len(lista_palavras))

O número de palavras é 403106


# Processo de normalização

### Retirando palavras repetidas

In [16]:
# função que recebe a 'lista_palavras', trasnforma a palavra para que fique toda minúscula 
def normalizacao(lista_palavras):
    lista_normalizada = []
    for palavra in lista_palavras:
        lista_normalizada.append(palavra.lower())
    return lista_normalizada

In [17]:
lista_normalizada = normalizacao(lista_palavras)
lista_normalizada[:500]

['imagem',
 'temos',
 'a',
 'seguinte',
 'classe',
 'que',
 'representa',
 'um',
 'usuário',
 'no',
 'nosso',
 'sistema',
 'java',
 'para',
 'salvar',
 'um',
 'novo',
 'usuário',
 'várias',
 'validações',
 'são',
 'feitas',
 'como',
 'por',
 'exemplo',
 'ver',
 'se',
 'o',
 'nome',
 'só',
 'contém',
 'letras',
 'o',
 'cpf',
 'só',
 'números',
 'e',
 'ver',
 'se',
 'o',
 'usuário',
 'possui',
 'no',
 'mínimo',
 'anos',
 'veja',
 'o',
 'método',
 'que',
 'faz',
 'essa',
 'validação',
 'java',
 'suponha',
 'agora',
 'que',
 'eu',
 'tenha',
 'outra',
 'classe',
 'a',
 'classe',
 'produto',
 'que',
 'contém',
 'um',
 'atributo',
 'nome',
 'e',
 'eu',
 'quero',
 'fazer',
 'a',
 'mesma',
 'validação',
 'que',
 'fiz',
 'para',
 'o',
 'nome',
 'do',
 'usuário',
 'ver',
 'se',
 'só',
 'contém',
 'letras',
 'e',
 'aí',
 'vou',
 'criar',
 'outro',
 'método',
 'para',
 'fazer',
 'a',
 'mesma',
 'validação',
 'ou',
 'criar',
 'uma',
 'interface',
 'ou',
 'uma',
 'classe',
 'que',
 'tanto',
 'usuario

### Texto normalizado, todas as palavras que contém apenas letras foram separadas em uma lista, e trasnformada em minúscula.

In [18]:
# A funcção set, retorna os valores da lista sem repetições
set([4, 4, 0, 1, 2, 3, 3, 3, 5, 6, 6, 8, 9, 7])

{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}

In [19]:
len(set(lista_normalizada))

18465

### De um total de 403106 palavras, fora reduzido para 18465 palavras sem reperições e sem caracteres

In [25]:
''' Função definida para receber uma pavara e gerar um fatiamento no lado esquedo e no lado direito da palavra,
    para que a letra de teste seja alocada entre elas e passando de posição em posição '''

palavra_exemplo = "lgica"
def gerador_palavras(palavra):
    fatias = []
    for i in range(len(palavra) + 1):
        fatias.append((palavra[:i], palavra[i:]))
    print(fatias)
    #palavras_geradas = insere_letras(fatias)
    #return palavras_geradas

In [26]:
''' Exemplo do que irá ocorrer com a função em cada palavra, de início, toda a palavra estará no canto direito
    e a letra será inserida na posição 1, então cada letra da palavra se deslocara para a esquerda e a letra
    de complementação se encaixara em cada espaço dessa transição'''
gerador_palavras(palavra_exemplo)

[('', 'lgica'), ('l', 'gica'), ('lg', 'ica'), ('lgi', 'ca'), ('lgic', 'a'), ('lgica', '')]


In [29]:
''' Função que insere uma nova letra, percorrendo todas as letras da variável destacada, e concatenando com o resultado
    da função 'gerador_de_palavras' '''
palavra_exemplo = "lgica"
def insere_letras(fatias):
    novas_palavras = []
    letras = 'abcdefghijklmnopqrstuvwxyzàáâãèéêìíîòóôõùúûç'
    for E,  D in fatias:    
        for letra in letras:  
            novas_palavras.append(E + letra + D)
    return novas_palavras


''' Função que fatia as palavras'''
def gerador_palavras(palavra):
    fatias = []
    for i in range(len(palavra)):
        fatias.append((palavra[:i], palavra[i:]))
    palavras_geradas = insere_letras(fatias)
    return palavras_geradas

palavras_geradas = gerador_palavras(palavra_exemplo)
print(palavras_geradas)

['algica', 'blgica', 'clgica', 'dlgica', 'elgica', 'flgica', 'glgica', 'hlgica', 'ilgica', 'jlgica', 'klgica', 'llgica', 'mlgica', 'nlgica', 'olgica', 'plgica', 'qlgica', 'rlgica', 'slgica', 'tlgica', 'ulgica', 'vlgica', 'wlgica', 'xlgica', 'ylgica', 'zlgica', 'àlgica', 'álgica', 'âlgica', 'ãlgica', 'èlgica', 'élgica', 'êlgica', 'ìlgica', 'ílgica', 'îlgica', 'òlgica', 'ólgica', 'ôlgica', 'õlgica', 'ùlgica', 'úlgica', 'ûlgica', 'çlgica', 'lagica', 'lbgica', 'lcgica', 'ldgica', 'legica', 'lfgica', 'lggica', 'lhgica', 'ligica', 'ljgica', 'lkgica', 'llgica', 'lmgica', 'lngica', 'logica', 'lpgica', 'lqgica', 'lrgica', 'lsgica', 'ltgica', 'lugica', 'lvgica', 'lwgica', 'lxgica', 'lygica', 'lzgica', 'làgica', 'lágica', 'lâgica', 'lãgica', 'lègica', 'légica', 'lêgica', 'lìgica', 'lígica', 'lîgica', 'lògica', 'lógica', 'lôgica', 'lõgica', 'lùgica', 'lúgica', 'lûgica', 'lçgica', 'lgaica', 'lgbica', 'lgcica', 'lgdica', 'lgeica', 'lgfica', 'lggica', 'lghica', 'lgiica', 'lgjica', 'lgkica', 'lglica',

# Corretor

In [41]:
# Retorna a palavra que tem maior probabilidade de ser a correta em relaçao ao gerador de palavras e a base 'artigos'
def corretor(palavra):
    palavras_geradas = gerador_palavras(palavra)
    palavra_correta = max(palavras_geradas, key = probabilidade)
    return palavra_correta

In [37]:
# retorna as palavras com maiores frequencias em 'artigos'
frequencia = nltk.FreqDist(lista_normalizada)
total_palavras = len(lista_normalizada)
frequencia.most_common(10)

[('de', 15502),
 ('o', 14056),
 ('que', 12230),
 ('a', 11099),
 ('e', 10501),
 ('para', 7710),
 ('um', 6368),
 ('é', 5899),
 ('uma', 5220),
 ('do', 5124)]

In [38]:
# retorna a probabilidade a palavra indicada, em relação a quantidade de palavras em 'artigos'
def probabilidade(palavra_gerada):
    return frequencia[palavra_gerada]/total_palavras

probabilidade("lógica")

0.00023815075935361914

In [50]:
# Resultado esperado com as funções, retorna todas as possível palavras, e retorna como resultado a com maior probabilidade
corretor('lóica')

['alóica', 'blóica', 'clóica', 'dlóica', 'elóica', 'flóica', 'glóica', 'hlóica', 'ilóica', 'jlóica', 'klóica', 'llóica', 'mlóica', 'nlóica', 'olóica', 'plóica', 'qlóica', 'rlóica', 'slóica', 'tlóica', 'ulóica', 'vlóica', 'wlóica', 'xlóica', 'ylóica', 'zlóica', 'àlóica', 'álóica', 'âlóica', 'ãlóica', 'èlóica', 'élóica', 'êlóica', 'ìlóica', 'ílóica', 'îlóica', 'òlóica', 'ólóica', 'ôlóica', 'õlóica', 'ùlóica', 'úlóica', 'ûlóica', 'çlóica', 'laóica', 'lbóica', 'lcóica', 'ldóica', 'leóica', 'lfóica', 'lgóica', 'lhóica', 'lióica', 'ljóica', 'lkóica', 'llóica', 'lmóica', 'lnóica', 'loóica', 'lpóica', 'lqóica', 'lróica', 'lsóica', 'ltóica', 'luóica', 'lvóica', 'lwóica', 'lxóica', 'lyóica', 'lzóica', 'làóica', 'láóica', 'lâóica', 'lãóica', 'lèóica', 'léóica', 'lêóica', 'lìóica', 'líóica', 'lîóica', 'lòóica', 'lóóica', 'lôóica', 'lõóica', 'lùóica', 'lúóica', 'lûóica', 'lçóica', 'lóaica', 'lóbica', 'lócica', 'lódica', 'lóeica', 'lófica', 'lógica', 'lóhica', 'lóiica', 'lójica', 'lókica', 'lólica',

'lógica'

In [51]:
# Resultado esperado com as funções, retorna todas as possível palavras, e retorna como resultado a com maior probabilidade
corretor('hoe')

['ahoe', 'bhoe', 'choe', 'dhoe', 'ehoe', 'fhoe', 'ghoe', 'hhoe', 'ihoe', 'jhoe', 'khoe', 'lhoe', 'mhoe', 'nhoe', 'ohoe', 'phoe', 'qhoe', 'rhoe', 'shoe', 'thoe', 'uhoe', 'vhoe', 'whoe', 'xhoe', 'yhoe', 'zhoe', 'àhoe', 'áhoe', 'âhoe', 'ãhoe', 'èhoe', 'éhoe', 'êhoe', 'ìhoe', 'íhoe', 'îhoe', 'òhoe', 'óhoe', 'ôhoe', 'õhoe', 'ùhoe', 'úhoe', 'ûhoe', 'çhoe', 'haoe', 'hboe', 'hcoe', 'hdoe', 'heoe', 'hfoe', 'hgoe', 'hhoe', 'hioe', 'hjoe', 'hkoe', 'hloe', 'hmoe', 'hnoe', 'hooe', 'hpoe', 'hqoe', 'hroe', 'hsoe', 'htoe', 'huoe', 'hvoe', 'hwoe', 'hxoe', 'hyoe', 'hzoe', 'hàoe', 'háoe', 'hâoe', 'hãoe', 'hèoe', 'héoe', 'hêoe', 'hìoe', 'híoe', 'hîoe', 'hòoe', 'hóoe', 'hôoe', 'hõoe', 'hùoe', 'húoe', 'hûoe', 'hçoe', 'hoae', 'hobe', 'hoce', 'hode', 'hoee', 'hofe', 'hoge', 'hohe', 'hoie', 'hoje', 'hoke', 'hole', 'home', 'hone', 'hooe', 'hope', 'hoqe', 'hore', 'hose', 'hote', 'houe', 'hove', 'howe', 'hoxe', 'hoye', 'hoze', 'hoàe', 'hoáe', 'hoâe', 'hoãe', 'hoèe', 'hoée', 'hoêe', 'hoìe', 'hoíe', 'hoîe', 'hoòe',

'hoje'

### Este corretor funciona apenas quando a palavra a ser corrigida está faltando uma letra e as outras estão corretas, não funciona caso a palavra tenha uma letra  a mais, ou se está faltando uma letra e no meio da palavra ainda existe outra letra errada como por exemplo a falta de um acento, e vale ressaltar que são probabilidades, o que não existe certeza absoluta da palavra retornada pelo corretor ser a esperada.

# Avaliador

In [68]:
def cria_dados_teste(nome_arquivo):
    lista_palavras_teste = []
    f = open(nome_arquivo, "r", errors = 'ignore', encoding = 'utf-8')
    for linha in f:
        correta, errada = linha.split()
        lista_palavras_teste.append((correta, errada))
    f.close()
    return lista_palavras_teste

lista_teste = cria_dados_teste('corretor-master/palavras.txt')

In [70]:
#print(lista_teste)

In [85]:
    
def avaliador(testes):
    numero_palavras = len(testes)
    acertou = 0
    for correta, errada in testes:
        palavra_corrigida = corretor(errada)
        if palavra_corrigida == correta:
            acertou += 1
    taxa_acerto = round(acertou*100/numero_palavras, 2)
    print(f"{taxa_acerto}% de {numero_palavras} palavras")

In [86]:
avaliador(lista_teste)

['apyodemos', 'bpyodemos', 'cpyodemos', 'dpyodemos', 'epyodemos', 'fpyodemos', 'gpyodemos', 'hpyodemos', 'ipyodemos', 'jpyodemos', 'kpyodemos', 'lpyodemos', 'mpyodemos', 'npyodemos', 'opyodemos', 'ppyodemos', 'qpyodemos', 'rpyodemos', 'spyodemos', 'tpyodemos', 'upyodemos', 'vpyodemos', 'wpyodemos', 'xpyodemos', 'ypyodemos', 'zpyodemos', 'àpyodemos', 'ápyodemos', 'âpyodemos', 'ãpyodemos', 'èpyodemos', 'épyodemos', 'êpyodemos', 'ìpyodemos', 'ípyodemos', 'îpyodemos', 'òpyodemos', 'ópyodemos', 'ôpyodemos', 'õpyodemos', 'ùpyodemos', 'úpyodemos', 'ûpyodemos', 'çpyodemos', 'payodemos', 'pbyodemos', 'pcyodemos', 'pdyodemos', 'peyodemos', 'pfyodemos', 'pgyodemos', 'phyodemos', 'piyodemos', 'pjyodemos', 'pkyodemos', 'plyodemos', 'pmyodemos', 'pnyodemos', 'poyodemos', 'ppyodemos', 'pqyodemos', 'pryodemos', 'psyodemos', 'ptyodemos', 'puyodemos', 'pvyodemos', 'pwyodemos', 'pxyodemos', 'pyyodemos', 'pzyodemos', 'pàyodemos', 'páyodemos', 'pâyodemos', 'pãyodemos', 'pèyodemos', 'péyodemos', 'pêyodemos'